packages/web-forms/src/components/common/map/useMapBlock.ts

```diff
@@ -0,0 +1,307 @@
+import {
+	getSavedStyles,
+	getSelectedStyles,
+	getUnselectedStyles,
+} from '@/components/common/map/map-styles.ts';
+import type { FeatureCollection } from 'geojson';
+import { Map, MapBrowserEvent, View } from 'ol';
+import { Zoom } from 'ol/control';
+import { getCenter } from 'ol/extent';
+import Feature from 'ol/Feature';
+import GeoJSON from 'ol/format/GeoJSON';
+import { LineString, Point, Polygon } from 'ol/geom';
+import TileLayer from 'ol/layer/Tile';
+import WebGLVectorLayer from 'ol/layer/WebGLVector';
+import type { Pixel } from 'ol/pixel';
+import { fromLonLat } from 'ol/proj';
+import { OSM } from 'ol/source';
+import VectorSource from 'ol/source/Vector';
+import { computed, shallowRef, watch } from 'vue';
+import { get as getProjection } from 'ol/proj';
+
+type GeometryType = LineString | Point | Polygon;
+
+const STATES = {
+	LOADING: 'loading',
+	READY: 'ready',
+	ERROR: 'error',
+} as const;
+
+const DEFAULT_GEOJSON_PROJECTION = 'EPSG:4326';
+const DEFAULT_VIEW_PROJECTION = 'EPSG:3857';
+const DEFAULT_VIEW_CENTER = [0, 0];
+const MAX_ZOOM = 16;
+const MIN_ZOOM = 2;
+const GEOLOCATION_TIMEOUT_MS = 10 * 1000;
+const ANIMATION_TIME = 1000;
+const SMALL_DEVICE_WIDTH = 576;
+const FEATURE_ID_PROPERTY = 'odk_feature_id';
+const SAVED_ID_PROPERTY = 'savedId';
+const SELECTED_ID_PROPERTY = 'selectedId';
+
+export function useMapBlock() {
+	const currentState = shallowRef<(typeof STATES)[keyof typeof STATES]>(STATES.LOADING);
+	const errorMessage = shallowRef<{ title: string; message: string } | undefined>();
+	let mapInstance: Map | undefined;
+	const savedFeature = shallowRef<Feature<GeometryType> | undefined>();
+	const selectedFeature = shallowRef<Feature<GeometryType> | undefined>();
+	const selectedFeatureProperties = computed(() => {
+		return selectedFeature.value?.getProperties();
+	});
+
+	const featuresSource = new VectorSource();
+	const featuresVectorLayer = new WebGLVectorLayer({
+		source: featuresSource,
+		style: [
+			...getUnselectedStyles(FEATURE_ID_PROPERTY, SELECTED_ID_PROPERTY, SAVED_ID_PROPERTY),
+			...getSelectedStyles(FEATURE_ID_PROPERTY, SELECTED_ID_PROPERTY, SAVED_ID_PROPERTY),
+			...getSavedStyles(FEATURE_ID_PROPERTY, SAVED_ID_PROPERTY),
+		],
+		variables: { [SAVED_ID_PROPERTY]: '', [SELECTED_ID_PROPERTY]: '' },
+	});
+
+	const initializeMap = (mapContainer: HTMLElement, geoJSON: FeatureCollection): void => {
+		if (mapInstance) {
+			return;
+		}
+
+		mapInstance = new Map({
+			target: mapContainer,
+			layers: [new TileLayer({ source: new OSM() }), featuresVectorLayer],
+			view: new View({
+				center: DEFAULT_VIEW_CENTER,
+				zoom: MIN_ZOOM,
+				// Prevent map cloning at low zoom during panning, which disrupts feature selection.
+				multiWorld: false,
+				projection: DEFAULT_VIEW_PROJECTION,
+				extent: getProjection(DEFAULT_VIEW_PROJECTION)?.getExtent(),
```

Comment:
@matthew-white You will need to add these lines

@srujner found an issue ([video](https://jam.dev/c/4807b3dd-a7f9-477e-be95-ddec5c6a36d9)) where the selection of features was disrupted when panning clone the map. You can see the fix in [my test server](https://effortless-dragon-d2ff5f.netlify.app/).